solidity/contracts/bridge/Bridge.sol

```diff
@@ -794,10 +949,574 @@ contract Bridge is Ownable {
+                        keccak256(
+                            abi.encodePacked(walletPubKeyHash, walletPubKeyHash)
+                        )
+                    )
+                ],
+                "Redemption transaction contains single change output"
+            );
+        }
+
+        return info;
+    }
+
+    // TODO: Function `notifyRedemptionTimeout. That function must:
+    //       1. Take a the `walletPubKey` and `redeemerOutputHash` as params.
+    //       2. Build the redemption key using those params.
+    //       3. Use the redemption key and take the request from
+    //          `pendingRedemptions` mapping.
+    //       4. If request doesn't exist in mapping - revert.
+    //       5. If request exits, and is timed out - remove the redemption key
+    //          from `pendingRedemptions` and put it to `invalidRedemptions`.
+    //          No need to check if `invalidRedemptions` mapping already contains
+    //          that key because `requestRedemption` blocks requests targeting
+    //          non-active wallets so there is no possibility that the given
+    //          redemption key could be reported as timed out multiple times.
+    //          On the other hand, if given redemption key was already
+    //          marked as invalid due to an amount-related fraud, it will not
+    //          be possible to report a time out on it since it won't be
+    //          present in `pendingRedemptions` mapping.
+    //       6. Return the `requestedAmount` to the `redeemer`.
+    //       7. Reduce the `pendingRedemptionsValue` (`wallets` mapping) for
+    //          given wallet by request's redeemable amount computed as
+    //          `requestedAmount - treasuryFee`.
+    //       8. Punish the wallet, probably by slashing its operators.
+    //       9. Change wallet's state in `wallets` mapping to `MovingFunds` in
+    //          order to prevent against new redemption requests hitting
+    //          that wallet.
+    //      10. Expect the wallet to transfer its funds to another healthy
+    //          wallet (just as in case of failed heartbeat).
+
+    // TODO: Function `submitRedemptionFraudProof`. That function must:
+    //       1. Take a `BitcoinTx.Info` and `BitcoinTx.Proof` of the
+    //          fraudulent transaction. It should also accept `walletPubKeyHash`.
+    //       2. Perform SPV proof to make sure it occurred on Bitcoin chain.
+    //          If not - revert.
+    //       3. Check if transaction hash is present in `fraudulentTransactions`
+    //          mapping. If present - revert.
+    //       4. Validate the input vector contains the main UTXO (fraudulent
+    //          transaction can use multiple inputs deliberately) for given
+    //          `walletPubKeyHash`. If invalid - revert. An open question is
+    //          what to do if transaction has proper outputs but wrong number
+    //          of inputs and it can't be processed by `submitRedemptionProof`.
```

Comment:
Yeah, this is an alternative but still an implementation detail that can be decided during the actual implementation. Also, keep in mind this approach still doesn't protect very large transactions since we have to do SPV proof (or sighash preimage processing alternatively) anyway.

Reply:
What would be problematic during SPV proof validation? Large hash or validations of bytes that needs to be done during SPV proof? Pointing to a specific input/output may not solve all problems but at least let us avoid gas limit problems when having to iterate and find problematic inputs/outputs.

I am spending so much time on these TODOs because I think they have to be precise enough that the auditors can look at them as well. We also need to be sure that we will not have to change anything in the audited sweep proof function when implementing fraud proofs and timeouts. Or at least those changes will be cosmetic.

Reply:
That change should close this thread: https://github.com/keep-network/tbtc-v2/commit/dcea8eb4c4194ed726c589894ea0718e275503a0